src/matcouply/penalties.py

```diff
@@ -1322,3 +1322,98 @@ def penalty(self, x):
+
+    More information as well as experiments can be found in :cite:p:`chatzis2023timeaware`.
+
+    **The proximal operator**
+
+    Taking the partial derivative of the objective function with respect to each auxiliary
+    variable results in a tridiagonal system of equations. More specifically:
+
+    .. math::
+
+    \begin{flalign}
+        & \bigl( 2\lambda_{\M{B}} + \rho_{\M{B}_1}\bigr) \M{Z}_{\M{B}_1} -2\lambda_{\M{B}}\M{Z}_{\M{B}_2} = \rho_{\M{B}_1} \bigr( \M{B}_1 + \M{\mu}_{\M{B}_1} \bigl) \notag \\
+        & \bigl(4\lambda_{\M{B}}+\rho_{\M{B}_k}\bigr)\M{Z}_{\M{B}_k}-2\lambda_{\M{B}} \bigl(\M{Z}_{\M{B}_{k-1}}+\M{Z}_{\M{B}_{k+1}} \bigr) =\rho_{\M{B}_k} \bigr(\M{B}_k+\M{\mu}_{\M{B}_k}\bigl), \qquad k = 2,\ldots,K-1 \notag \\
+        & \vdots \notag \\
+        & \bigl( 2\lambda_{\M{B}} + \rho_{\M{B}_K}\bigr) \M{Z}_{\M{B}_K} -2\lambda_{\M{B}}\M{Z}_{\M{B}_{K-1}} = \rho_{\M{B}_K} \bigr( \M{B}_K + \,{\mu}_{\M{B}_K} \bigl) \notag
+    \end{flalign}
+
+    We then form a tridiagonal matrix :math:`\M{A}` and solve the system of equations using Thomas algorithm.
+
+    Keep in mind that once we apply a norm-based penalty in the factor matrices, we need to also penalize the
+    norm of all other factor matrices :cite:p:`roald2021parafac2` to overcome the scaling ambiguity of PARAFAC2.
+    This can be done, for example, by adding ridge penalties on the norms of the other two factor matrices.
+
+    Parameters
+    ----------
+    smoothness_l : float
+        Hyperparemeter controlling the strength of the smoothness penalty.
+    aux_init : {"random_uniform", "random_standard_normal", "zeros", tl.tensor(ndim=2), list of tl.tensor(ndim=2)}
+        Initialisation method for the auxiliary variables
+    dual_init : {"random_uniform", "random_standard_normal", "zeros", tl.tensor(ndim=2), list of tl.tensor(ndim=2)}
+        Initialisation method for the auxiliary variables
+    """
+
+    def __init__(self, smoothness_l, aux_init="random_uniform", dual_init="random_uniform"):
+        super().__init__(aux_init=aux_init, dual_init=dual_init)
+        self.smoothness_l = smoothness_l
+
+    @copy_ancestor_docstring
+    def factor_matrices_update(self, factor_matrices, feasibility_penalties, auxes):
+
+        B_is = factor_matrices
+        rhos = feasibility_penalties
+
+        rhs = [rhos[i] * factor_matrices[i] for i in range(len(B_is))]
+
+        # Construct matrix A to peform thomas algorithm on
+
+        A = np.zeros((len(B_is) , len(B_is)))
+
+        for i in range(len(B_is)):
+            for j in range(len(B_is)):
+                if i == j:
+                    A[i , j] = 4 * self.smoothness_l + rhos[i]
+                elif (i == j - 1) or (i == j + 1):
+                    A[i , j] = - 2 * self.smoothness_l
+                else:
+                    pass
+
+        A[0, 0] -= 2 * self.smoothness_l
+        A[len(B_is) - 1 , len(B_is) - 1] -= 2 * self.smoothness_l
```

Comment:
```suggestion
     A = (
        tl.diag([self._get_laplace_coef(i) + rho for i, rho in enumerate(rhos)], k=0)
        - tl.diag(tl.ones(I-1) * 2 * self.smoothness_l, k=1)
        - tl.diag(tl.ones(I-1) * 2 * self.smoothness_l, k=-1)
    )
```

This will reduce the assembly complexity from O(I^2) to O(I). We'd need the `_get_laplace_coef` method though, which could look like this for example:

```python
def _get_laplace_coef(self, i, feasibility_penalties):
    I = len(feasibility_penalties)
    rhos = feasibility_penalties
    if i == 0 or i = I - 1:
        return 2 * self.smoothness_l + rhos[i]
    return 4 * self.smoothness_l + rhos[i]
```

Reply:
This was addressed in 95877352ccbba80b65ed9eaa865d55f68814dd68.